### PR TITLE
* add first version of tutorials

### PR DESCRIPTION
**Pull request by benjaminLedel**
bug #8 

Added the first version of plugin creation. Happy to receive some feedback.

### CODE REVIEW

File: tutorial_plugin_java.md [hunk 1]
```diff
@@ -0,0 +1,154 @@
+﻿---
+layout: page
+title: Tutorial Plugin Java
+permalink: /plugin/tutorial/java
+---
+
+# How to create a plugin for SmartSHARK
```
Comment (sherbold) [inline tutorial_plugin_java.md]: "This repository" does not work here.

Maybe "This tutorial exaplins how to create Jjava plugin for SmartSHARK"?

File: tutorial_plugin_java.md [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+This repository contains a minimal working example of a java plugin for SmartSHARK. This short readme is a tutorial on how to program a plugin starting from the ground. The first section will provide you with the basic information. More advanced information will be outlined for each file in the files section.
+
+## Basics
```
Comment (sherbold) [inline tutorial_plugin_java.md]: This section is not needed, there are no other subsections. Just remove this and start directly with "## Get started"

File: tutorial_plugin_java.md [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+In this section, we will shortly explain the basics of plugin development.
+### Get started
+This  is a quick start guide.
```
Comment (sherbold) [inline tutorial_plugin_java.md]: This intro sentence can be dropped

File: tutorial_plugin_java.md [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+---
+
+# How to create a plugin for SmartSHARK
+This repository contains a minimal working example of a java plugin for SmartSHARK. This short readme is a tutorial on how to program a plugin starting from the ground. The first section will provide you with the basic information. More advanced information will be outlined for each file in the files section.
+
+## Basics
+In this section, we will shortly explain the basics of plugin development.
+### Get started
+This  is a quick start guide. 
+1. Modify the content of the setup.py and the info.json with your information
```
Comment (sherbold) [inline tutorial_plugin_java.md]: There is no setup.py for Java plugins. My guess is you mean the other files in the plugin_packaging?

File: tutorial_plugin_java.md [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+2. Create a new folder, with the name of the plugin
+3. This folder is your workspace. Add any java files here.
```
Comment (sherbold) [inline tutorial_plugin_java.md]: Really? This sounds pythonic.

File: tutorial_plugin_java.md [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+4. The main.py contains the main method and start point of the plugin. It automatically setups some logging parameters and start the timing of the execution.
```
Comment (sherbold) [inline tutorial_plugin_java.md]: Again, python.

File: tutorial_plugin_java.md [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+## Basics
+In this section, we will shortly explain the basics of plugin development.
+### Get started
+This  is a quick start guide. 
+1. Modify the content of the setup.py and the info.json with your information
+2. Create a new folder, with the name of the plugin
+3. This folder is your workspace. Add any java files here.
+4. The main.py contains the main method and start point of the plugin. It automatically setups some logging parameters and start the timing of the execution. 
+5. Parse additional parameter and start your plugin
```
Comment (sherbold) [inline tutorial_plugin_java.md]: parameters

File: tutorial_plugin_java.md [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+### Parameter
```
Comment (sherbold) [inline tutorial_plugin_java.md]: Should be "Command Line Parameters"
pull up section

File: tutorial_plugin_java.md [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+#### Get a parameter
```
Comment (sherbold) [inline tutorial_plugin_java.md]: pull up section

File: tutorial_plugin_java.md [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+All parameters are automatically parsed in the main method.
+Example:
+
+      Parameter param = Parameter.getInstance();
+	  param.init(args);
+
```
Comment (sherbold) [inline tutorial_plugin_java.md]: use markdown for all code:
```
Parameter param = Parameter.getInstance();
param.init(args);
```

File: tutorial_plugin_java.md [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+ #### Add a parameter
+ To add a parameter follow these steps:
+ 1. Add the parameter to the info.json
+ 2. Add the parameter to the execute.sh in the plugin_packaging folder. Add the parameter to COMMAND and decide, if the parameter is optional (Be careful that you have the same position as in your info.json)
+ 3. Read your parameter in the java main method
+ ### Database
```
Comment (sherbold) [inline tutorial_plugin_java.md]: Should be "Reading and Writing Data"
pull up section

File: tutorial_plugin_java.md [hunk 12]
```diff
@@ -0,0 +1,154 @@
...
+ #### Create a database connection
+It is possible to create a database connection with morphia library. The following code creates an uri from the connection arguments and connects to the database.
```
Comment (sherbold) [inline tutorial_plugin_java.md]: Link to Morphia
URI

File: tutorial_plugin_java.md [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+
+    datastore = morphia.createDatastore(
+						new MongoClient(Parameter.getInstance().getDbHostname(), Parameter.getInstance().getDbPort()),
+						Parameter.getInstance().getDbName());
+
```
Comment (sherbold) [inline tutorial_plugin_java.md]: Markdown for code

File: tutorial_plugin_java.md [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+ #### Modify the database
```
Comment (sherbold) [inline tutorial_plugin_java.md]: Writing data
Pull up section

Why is there no "Reading data" beforehand?

File: tutorial_plugin_java.md [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+    ],
+    
+    "desc": "",
+    
+    "collection_name": "file_action"
+    
+    }
+
+ ### Build and Installation
+ Execute the following commands to build the plugin and create an archive for uploading to the SmartShark Server.
```
Comment (sherbold) [inline tutorial_plugin_java.md]: link to the servershark

File: tutorial_plugin_java.md [hunk 16]
```diff
@@ -0,0 +1,154 @@
...
+ 
+    ./gradlew build
+    ./plugin-packaging/build_plugin.sh
+
+### Add external libraries
+You are able to add any external library to the build.gradle file in the root folder.
+Additional information can be found in the documentation of gradle [here](https://gradle.org/).
+
+## Files
```
Comment (sherbold) [inline tutorial_plugin_java.md]: Should be at the beginning

File: tutorial_plugin_java.md [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+In this section, we will provide detailed information about any file of the minimal working example. 
+### .travis.yml
```
Comment (sherbold) [inline tutorial_plugin_java.md]: Use an itemized list instead of sections.

File: tutorial_plugin_java.md [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+File for the Travis-Build System. 
+### JmweSHARK.java
+The java main class of the plugin. The main method is automatically called by execution of the plugin.
```
Comment (sherbold) [inline tutorial_plugin_java.md]: Java

File: tutorial_plugin_java.md [hunk 19]
```diff
@@ -0,0 +1,154 @@
...
+### build.gradle
+Additional information about the build.gradle file can be found in the documentation from gradle [here](https://docs.gradle.org/current/userguide/tutorial_using_tasks.html).
```
Comment (sherbold) [inline tutorial_plugin_java.md]: Start this with "The gradle build file for the project."

File: tutorial_plugin_java.md [hunk 20]
```diff
@@ -0,0 +1,154 @@
...
+### build_plugin.sh
+Script for creating a zip archive file of the plugin that is able to upload to the ServerSHARK.
```
Comment (sherbold) [inline tutorial_plugin_java.md]: "Script for creating an archive file of the plugin that can be used by the ServerSHARK."
(link ServerSHARK)

File: tutorial_plugin_java.md [hunk 21]
```diff
@@ -0,0 +1,154 @@
...
+### .travis.yml
+File for the Travis-Build System. 
+### JmweSHARK.java
+The java main class of the plugin. The main method is automatically called by execution of the plugin.
+### build.gradle
+Additional information about the build.gradle file can be found in the documentation from gradle [here](https://docs.gradle.org/current/userguide/tutorial_using_tasks.html).
+### build_plugin.sh
+Script for creating a zip archive file of the plugin that is able to upload to the ServerSHARK. 
+### execute.sh
+The plugin is execution starts with this script file. You may add here some additional parameters for your plugin.
```
Comment (sherbold) [inline tutorial_plugin_java.md]: The plugin is executed with this script.

File: tutorial_plugin_java.md [hunk 22]
```diff
@@ -0,0 +1,154 @@
...
+### info.json
+Contains important meta information about the plugin. This file is readed by the ServerSHARK to ensure that the requirements to execute the plugin are given. 
+### install.sh
+Install-Script: Usually, you do not need to change anything here. 
+### schema.json
+Contains any information about database changes that are possible while executing this plugin.
```
Comment (sherbold) [inline tutorial_plugin_java.md]: "Contains information about collections and fields that may be modified in the database by this plugin."

File: tutorial_plugin_python.md [hunk 23]
```diff
@@ -0,0 +1,157 @@
+﻿---
```
Comment (sherbold) [inline tutorial_plugin_python.md]: Most from above applies here as well. Will only check after the changes above are also integrated here.

**Comment by benjaminLedel** [comment]
I updated the java version. I will look into the python version tomorrow.

**Comment by benjaminLedel** [comment]
Ready for a second review

File: tutorial_plugin_java.md [hunk 24]
```diff
@@ -0,0 +1,160 @@
...
+All parameters are automatically parsed in the main method.
+Example:
+```
+      Parameter param = Parameter.getInstance();
+	  param.init(args);
+```
+### Add a parameter
+ To add a parameter follow these steps:
+ 1. Add the parameter to the info.json
+ 2. Add the parameter to the execute.sh in the plugin_packaging folder. Add the parameter to COMMAND and decide, if the parameter is optional (Be careful that you have the same position as in your info.json)
```
Comment (sherbold) [inline tutorial_plugin_java.md]: There is a dot missing and the last sentence should not be in brackets.

File: tutorial_plugin_java.md [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+ 3. Read your parameter in the Java main method
+## Reading and Writing Data
+### Create a database connection
+It is possible to create a database connection with [morphia](https://github.com/MorphiaOrg/morphia) library. The following code creates an uri from the connection arguments and connects to the database.
+```
+    datastore = morphia.createDatastore(
+						new MongoClient(Parameter.getInstance().getDbHostname(), Parameter.getInstance().getDbPort()),
+						Parameter.getInstance().getDbName());
+```
```
Comment (sherbold) [inline tutorial_plugin_java.md]: Check code formatting.

File: tutorial_plugin_java.md [hunk 26]
```diff
@@ -0,0 +1,160 @@
...
+3. The JmweShark.java is the main application of the Java plugin. It contains the Java main method and is therefore the start point. 
+4. In the example Java class, is some code to parse additional arguments (command line parameters) and setup the logging, the database connection and time measurements. 
+## Command Line Parameters
+### Get a parameter
+All parameters are automatically parsed in the main method.
+Example:
+```
+      Parameter param = Parameter.getInstance();
+	  param.init(args);
+```
```
Comment (sherbold) [inline tutorial_plugin_java.md]: Check code formatting

File: tutorial_plugin_java.md (hunk 25 above)
Comment (sherbold) [inline tutorial_plugin_java.md]: Check code formatting

File: tutorial_plugin_java.md [hunk 27]
```diff
@@ -0,0 +1,160 @@
...
+						Parameter.getInstance().getDbName());
+```
+### Reading data 
+You are able to read data from the database with an active database connection. Just use the [morphia](https://github.com/MorphiaOrg/morphia) syntax. For example:
+```
+    Query<Project> projects = datastore.createQuery(Project.class);
+    for (Project project : projects) {
+	    System.out.println(project.getName());
+	}
+```
```
Comment (sherbold) [inline tutorial_plugin_java.md]: Check code formatting

File: tutorial_plugin_java.md [hunk 28]
```diff
@@ -0,0 +1,160 @@
...
+    }
+    
+    ],
+    
+    "desc": "",
+    
+    "collection_name": "file_action"
+    
+    }
+	```
```
Comment (sherbold) [inline tutorial_plugin_java.md]: Check code formatting

File: tutorial_plugin_java.md [hunk 29]
```diff
@@ -0,0 +1,160 @@
...
+
+## Build and Installation
+Execute the following commands to build the plugin and create an archive for uploading to the [ServerSHARK](https://github.com/smartshark/serverSHARK) Server.
+  ```
+    ./gradlew build
+    ./plugin-packaging/build_plugin.sh
+```
```
Comment (sherbold) [inline tutorial_plugin_java.md]: Check code formatting

File: tutorial_plugin_python.md (hunk 23 above)
Comment (sherbold) [reply tutorial_plugin_python.md]: Looks mostly good, same as above. But also formatting issues.

File: tutorial_plugin_java.md [hunk 30]
```diff
@@ -0,0 +1,160 @@
+﻿---
+layout: page
+title: Tutorial Plugin Java
+permalink: /plugin/tutorial/java
+---
+
+# How to create a plugin for SmartSHARK
+This tutorial explains how to create a Java plugin for SmartSHARK. This short readme is a tutorial on how to program a plugin starting from the ground. The first section will provide you with the basic information. More advanced information will be outlined for each file in the files section.
```
Comment (sherbold) [inline tutorial_plugin_java.md]: Second sentence: "We explain how you can create a simple plugin for SmartSHARK using the minimal working example [jmweSHARK](LINK HERE).

**Comment by sherbold** [comment]
Second round of reviews done. Basically the code must be formatted correctly (starts at the beginning of the line, no tabs for indents within the README) and links to the minimal working examples are missing.

**Comment by benjaminLedel** [comment]
Please review again

**Comment by sherbold** [comment]
LGTM

**Pull request merged**